File: leetcode/problems_list/49-GroupAnagrams/solution.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
// Time Complexity: O(n * mlogm) where n is the number of strings and m is the length of the longest string
// Space Complexity: O(n) where n is the number of strings
var groupAnagrams = function (strs) {
  let anagramGroup = new Map(),
    arrayAnagramGroup = [];
  for (const curr of strs) {
    let sortedString = [...curr].sort().join("");
    anagramGroup[sortedString] = anagramGroup.has(sortedString)
      ? [...anagramGroup[sortedString], curr]
      : [curr];
  }
  for (const curr in anagramGroup) {
    arrayAnagramGroup.push(anagramGroup[curr]);
  }

  return arrayAnagramGroup;
};

/** Solition 2 Using Map
 *
 * var groupAnagrams = function (strs) {
 * const anagramGroup = new Map()
 * for (const curr of strs) {
 *  let sortedString = [...curr].sort().join("")
 *  if(!anagramGroup.has(sortedString)){
 *    anagramGroup.set(sortedString, [])
 *  }
 *  anagramGroup.get(sortedString).push(curr)
 * }
 *  return Array.from(anagramGroup.values())
 * };
 */

File: leetcode/problems_list/49-GroupAnagrams/solution.py
"""
Problem: Anagrams Grouping
Source: https://leetcode.com/problems/group-anagrams/
Time Complexity: O(n * m * log(m))
Space Complexity: O(n)

"""


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        grouped_anagrams = {}
        ans = []
        for curr in strs:
            sorted_string = ''.join(sorted(curr))
            if sorted_string in grouped_anagrams:
                ans[grouped_anagrams.get(sorted_string)].append(curr)
            else:
                grouped_anagrams[sorted_string] = len(ans)
                ans.append([curr])
        return ans
